File: HouseRobber.java
public class HouseRobber {
    public static void main(String[] args) {
        int[] nums = new int[]{2, 7, 9, 3, 1};
        rob(nums);
    }
    public static int rob(int[] nums) {
        // if (nums.length == 0) {
        //   return 0;
        // }

        // int maxMoney = 0;
        // int maxRob = 0;
        // int[] dp = new int[nums.length];
        // for (int i = 0; i < nums.length; i++) {
        //   dp[i] = nums[i];
        //   maxRob = Math.max(maxRob, nums[i]);
        // }

        // for (int i = 0; i < nums.length; i++) {
        //   for (int j = 0; j < i - 1; j++) {
        //     maxMoney = Math.max(maxMoney, dp[j] + nums[i]);
        //     dp[i] = maxMoney;
        //   }
        // }

        // for (int i : dp) {
        //   System.out.print(i + " ");
        // }

        // return Math.max(maxMoney, maxRob);

        int rob1 = 0;
        int rob2 = 0;
        for (int i = 0; i < nums.length; i++) {
          int temp = Math.max(nums[i] + rob1, rob2);
          rob1 = rob2;
          rob2 = temp;
        }

        return rob2;
    }
}

File: HouseRobber.py
from typing import List

class HouseRobber:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])
        dp = []
        for i in range(len(nums)):
            dp.append(1)
        
        dp[0] = nums[0]
        dp[1] = max(nums[1], nums[0])
        for i in range(2, len(nums)):
            dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])
        
        return dp[len(nums) - 1]